BuchiAutomizer for concurrent programs

schuessf opened the pull request:
This pull request allows to use the `BuchiAutomizer` for concurrent programs. It is based on the work from pull request #584 and consists of the following changes:
*  It is now possible to analyse the termination of concurrent programs represented by finite automata (the "product automata") using the already exisiting `BuchiAutomatonCegarLoop`. This needed some minor changes in order to work as expected.
* This requires the petrification of the existing icfg with an increasing number of thread instances (similar to the trace abstraction). In order to detect insufficient thread instances, the petrification was slightly modified and now adds self-loops to the insufficient-thread-location
* Additionally `AbstractBuchiCegarLoop` was refactored. Now it deals with the `LassoCheck` and creates interpolant automata out of it (including the iterations over different construction styles), instead of the implementing class. This makes it easier to implement this class, since you can mostly use the automata-theoretic operations and should also reduce code duplication in the future.

schuessf committed:
Move construction of automaton for refineFinite to abstract class

schuessf committed:
Add skeleton for petri-net based cegar loop

schuessf committed:
Add header

Heizmann committed:
bugfix: add missing break statement

Heizmann committed:
use existing methods

Heizmann committed:
bugfix: quantifier alternation in DualJunctionSaa

Heizmann committed:
fix bug in error message

Heizmann committed:
add examples

Heizmann committed:
add/modify examples

Heizmann committed:
add scripts for SMT-COMP 2022

Heizmann committed:
add auxiliary constructor

Heizmann committed:
add examples that show problems with the live variable optimization

Heizmann committed:
change to arrray index-based loop acceleration

schuessf committed:
Allow termination analysis for concurrent programs
* Introduce setting for the automaton type (not used yet)
* Let BuchiAutomatonCegarLoop accept an abstraction of type INwaOutgoingLetterAndTransitionProvider
* Minor CC fixes

schuessf committed:
Use petrification (should be enhanced further)

schuessf committed:
Avoid unsafe cast to IMLPredicate

Heizmann committed:
add tests

Heizmann committed:
typo

schuessf committed:
Bugfix: Use petrified icfg

schuessf committed:
Fix more unsafe casts

schuessf committed:
Fix backtranslation

schuessf committed:
Workaround for another unsafe cast

schuessf committed:
Add example

schuessf committed:
Revert accidental commits

schuessf committed:
Throw exception to avoid unsound result

schuessf committed:
Read the automaton type from the settings (now using correct PLUGIN_ID)

schuessf committed:
Refactoring, reduce duplicate code

schuessf committed:
Refactoring: Move iteration over construction-styles to abstract class

This allows refineBuchi to be as expected with just a interpolant automaton

schuessf committed:
Remove outcommented code

schuessf committed:
Remove BuchiAutomatonCegarLoop (now just a trivial skeleton)

schuessf committed:
Extract method

schuessf committed:
Use translateProgramStateToString instead of translateExpressionToString (respects variable blacklist)

schuessf committed:
Try to increase thread-number if needed

schuessf committed:
Add identifier to RankVarConstructor to avoid "already defined" exception

schuessf committed:
Refactoring: Move complete construction to BuchiCegarLoopFactory

schuessf committed:
Move identifier to BuchiCegarLoopFactory

schuessf committed:
Minor changes

danieldietsch commented on trunk/source/BuchiAutomizer/src/de/uni_freiburg/informatik/ultimate/plugins/generator/buchiautomizer/BuchiInterpolantAutomatonBuilder.java:
```diff
@@ -0,0 +1,209 @@
+		} else {
+			result.addState(true, false, precondition);
+			for (int i = 0; i < stemInterpolants.length; i++) {
+				addState(stemInterpolants[i], result);
```
why do you call `addState` sometimes directly on the NWA and sometimes via the `addState` wrapper?

schuessf replied:
Not sure, this is just some old unchanged code that was moved to an extra class.

danieldietsch commented:
Can you check why the various checks failed? 

It will take some time to review a PR this large -- I try and do it piece-wise. 

In general, it would be nice to squash large PRs so that they dont have so many commits, e.g., combine all cc fixes into one, combine new examples into one, try and keep features together (i.e., if you have something that took a couple of commits with various local fixes and rewrites, it would be nice to have it in one commit).

schuessf commented:
> Can you check why the various checks failed?

The log [here](https://jenkins.sopranium.de/job/Ultimate/job/Ultimate/view/change-requests/job/PR-589/2/console) yields this exception:
`Unable to read repository at https://monteverdi.informatik.uni-freiburg.de/nexus/content/repositories/eclipse-4.17. https://monteverdi.informatik.uni-freiburg.de/nexus/content/repositories/eclipse-4.17 is not a valid repository location`
I am not sure, why it failed this morning, it did build successfully right after the creation of this PR.

> It will take some time to review a PR this large -- I try and do it piece-wise.
> 
> In general, it would be nice to squash large PRs so that they dont have so many commits, e.g., combine all cc fixes into one, combine new examples into one, try and keep features together (i.e., if you have something that took a couple of commits with various local fixes and rewrites, it would be nice to have it in one commit).

Yes, sorry, required more refactoring than I expected. Basically the changes in [BuchiCegarLoopFactory.java](https://github.com/ultimate-pa/ultimate/pull/589/files#diff-ebb8b35ee436d840b1d75dd4ed329a4ef1f8b823e7a2e1456896b91c6984b87a) cover the first point, [BuchiAutomizerObserver.java](https://github.com/ultimate-pa/ultimate/pull/589/files#diff-39347dab6222b2d1a3a0b956e0e406c3ae09e26c6783c80d76f077145ceabcc0) the second point and the rest (which is obviously a lot!) is just refactoring that should make things easier for future changes (i.e. the last point).

schuessf committed:
Update documentation

schuessf committed:
Remove outcommented code

schuessf committed:
Add comment

schuessf committed:
Restructure code

schuessf committed:
Some more restructuring

schuessf committed:
Remove unused fields

schuessf committed:
Update headers

schuessf committed:
Remove counterexample from RefineBuchi

schuessf committed:
Allow usage of petri-net LBE

schuessf committed:
Add settings (with/without LBE)

schuessf committed:
Add benchexec xml

maul-esel commented on trunk/source/BuchiAutomizer/src/de/uni_freiburg/informatik/ultimate/plugins/generator/buchiautomizer/BuchiAutomizerObserver.java:
```diff
@@ -131,41 +138,71 @@ public boolean process(final IElement root) throws IOException {
+						factory.constructCegarLoop(petrified, witnessAutomaton);
+				final Result currentResult = currentCegarLoop.runCegarLoop();
+				if (currentResult != Result.NONTERMINATING
+						|| !containsInsufficientThreadLocation(currentCegarLoop.getCounterexample(),
```
I'm not sure I understand this check. If the result is that the program terminates (1st disjunct true), why do we not increase the number of threads?

maul-esel replied:
Also, a less crucial comment: If a nonterminating execution contains an "insufficient threads" location, do we always need to increase the thread count? Particularly, if the location appears only in the stem?

**Edit:** Ah, I see `containsInsufficientThreadLocation` looks only at the loop. Perhaps rename the method to make that clear?

schuessf replied:
* I added a self loop to the insufficient thread loactions in order to find, if we need more thread instances. If the program terminates, we know that we did not reach this location.
* If the program reaches this location, it cannot leave it anymore (the only outgoing edge is a self loop). I could also add the stem to the check, it does not make a difference.

maul-esel replied:
>  If the program terminates, we know that we did not reach this location.

Ah ok, of course.

> If the program reaches this location, it cannot leave it anymore (the only outgoing edge is a self loop). I could also add the stem to the check, it does not make a difference.

Even if one thread reaches such a location, other threads may still move, right?

schuessf replied:
Yes, the other threads could still move, the self-loop may not even occur in the lasso-loop. But we still reached the insufficient-thread-location. This could be a bit imprecise in some (hopefully rare) cases (similar for `TraceAbstraction`), but should be still correct.

maul-esel replied:
Yes, I was thinking about this case. I'm (reasonably) convinced that the current solution is sound.

(We can have a look in the future if in such cases we want to modify the nonterminating counterexample before we output it, but let's leave it for now.)

schuessf replied:
Yes, I think modifying the counterexample is a good idea. First we want to remove all self-loops in the insufficient thread from it. We could safely do that, since it does not change the location. If the loop is empty then, we need to increase the number of thread instances, otherwise we found a "real" counterexample and can stop.

We probably also want to remove the edge leading to the insufficient thread. This edge should only occur in the stem. There we need to slightly modifiy the locations, but I guess that should not be an issue.

Removing these edges should also be good, since we don't have a backtranslation for them (I think).

maul-esel commented on trunk/source/BuchiAutomizer/src/de/uni_freiburg/informatik/ultimate/plugins/generator/buchiautomizer/BuchiAutomizerObserver.java:
```diff
@@ -324,9 +358,16 @@ private void reportLTLPropertyIsViolated(final AbstractBuchiCegarLoop<IcfgEdge,
+	}
+
+	private static IcfgLocation
+			getProgramPoint(final NestedLassoRun<? extends IIcfgTransition<?>, IPredicate> counterexample) {
```
Maybe make `honda` part of the name? The ctex contains many program points.

schuessf replied:
Good idea, I will change that.

maul-esel commented on trunk/source/BuchiAutomizer/src/de/uni_freiburg/informatik/ultimate/plugins/generator/buchiautomizer/BuchiAutomizerObserver.java:
```diff
@@ -324,9 +358,16 @@ private void reportLTLPropertyIsViolated(final AbstractBuchiCegarLoop<IcfgEdge,
+			getProgramPoint(final NestedLassoRun<? extends IIcfgTransition<?>, IPredicate> counterexample) {
+		final IPredicate pred = counterexample.getLoop().getStateAtPosition(0);
+		// Return any program point from the predicate
+		// There is not really a "correct" answer here for concurrent programs
```
What role does the program point play in the counterexample-results? Should these be changed to take an `IPredicate` (i.e., an `SPredicate` or `MLPredicate`) instead?

schuessf replied:
If the program does not terminate, we get a `NonterminatingLassoResult` that specifies the line, where the loop starts. For sequential programs this is just the loop head, but for concurrent programs there is no really such a thing, since there can be multiple concurrent loops. Therefore I just return any of those states contained in the predicate. This should be one loop head of the involved loops. But I just found out that this is not always true, so I will change that.

schuessf replied:
Changed it in this commit: 5a81a7ce46a14f1fbbb2ca9b259ca8214dbf3eac
Now there is a method `private static IcfgEdge getHondaAction(...)` and we use this edge as an identifier. This should output better line numbers (we use just the source of this edge) for concurrent programs and should not make a difference for sequential programs.

maul-esel replied:
:+1:

maul-esel commented on trunk/source/BuchiAutomizer/src/de/uni_freiburg/informatik/ultimate/plugins/generator/buchiautomizer/cegar/AbstractBuchiCegarLoop.java:
```diff
@@ -204,47 +242,75 @@ public AbstractBuchiCegarLoop(final IIcfg<?> icfg, final RankVarConstructor rank
+		mBiaConstructionStyleSequence =
+				baPref.getEnum(BuchiAutomizerPreferenceInitializer.LABEL_BIA_CONSTRUCTION_STRATEGY,
+						BuchiInterpolantAutomatonConstructionStrategy.class).getBiaConstrucionStyleSequence(baPref);
+		mAutomataMinimizationAfterFeasbilityBasedRefinement = baPref.getEnum(
```
typo: Feasbility

schuessf replied:
Ah, another old, yet unspotted typo 😄

maul-esel reviewed (commented):
I really like the idea of this PR.

I took a look, but can't give any guarantees (I'm unfamiliar with this part of the code). I left a few comments and some questions.

schuessf committed:
Fix typo

schuessf committed:
Use IcfgEdge as element in NonterminatingLassoResult
This should produce results similar to FixpointNonTerminationResult (and consistent), some refactoring necessary

schuessf committed:
Reduce visibility, rearrange fields
